### ISSUE TITLE
Pagination Bug whilst using `hasPreviousPage` returning true.

### ISSUE THREAD

**Issue by izzygld** [wp-graphql/wp-graphql#430]
### `hasPreviousPage` returning true.

**Pagination**: the way I'm doing it, depends on these 4 variables returned from the request: `hasNextPage`, `hasPreviousPage`, `startCursor` and `endCursor`. If `hasNextPage` is true then I use `endCursor` to retrieve the next posts. 

Problem is, that `hasPreviousPage` is returning as true one extra time.

_ggl Query:_
```const postsQuery = gql
    query GetAllPosts ($first: Int, $last: Int, $endCursor: String, $startCursor: String) {
      posts(first:$first last:$last after:$endCursor before:$startCursor) {
        edges {
          node {
            title
            id
            slug
            excerpt
    featuredImage {
              guid
            }
          }
          cursor
        }
      pageInfo {
        hasNextPage
        hasPreviousPage
        startCursor
        endCursor
        }
      }
    }
  ```
  
  
  
 _Data Functions:_

  ```export default {
  data () {
    return {
      hasNextPage:true,
      hasPreviousPage:false,
      first: null,
      last: null,
      startCursor: null,
      endCursor: null
    }  
  },
  async asyncData({ route, app }) {
    const client = app.apolloProvider.defaultClient

    return {
      posts: await client
        .query({
          query: postsQuery,
          variables: {
              first: 5
          }
        })
    }
  },
  methods: {
    async paginate(i){
      if (i < 0) {
        this.first = null
        this.last = 5
        this.startCursor = this.posts.data.posts.pageInfo.startCursor
        this.endCursor = null
      } else if (i > 0) {
        this.first = 5
        this.last = null
        this.startCursor = null
        this.endCursor = this.posts.data.posts.pageInfo.endCursor
      }

      await this.$apolloProvider.defaultClient
        .query({
          query: postsQuery,
          variables: {
            first: this.first,
            last: this.last,
            endCursor: this.endCursor,
            startCursor: this.startCursor
          }
        }).then((posts) => {
          this.posts = posts
          if (i < 0) {
            this.hasNextPage = true
            this.hasPreviousPage = posts.data.posts.pageInfo.hasPreviousPage
          } else {
            this.hasPreviousPage = true
            this.hasNextPage = posts.data.posts.pageInfo.hasNextPage
          }
        })
    }
  },
}``

**Comment by jasonbahl** [comment]
Thanks for the details!should help troubleshoot

**Comment by strootje** [comment]
Not sure if this is related, but look in the file:
https://github.com/wp-graphql/wp-graphql/blob/develop/src/Type/PostObject/Connection/PostObjectConnectionResolver.php#L235

If shows you the "hasNextPage" and "hasPreviousPage" are set.
* if 'first' is set: "hasPreviousPage" will always be false
* if 'last' is set: "hasNextPage" will always be false

I've been struggling with the WP_Query object but there doesn't seem any way to be able to properly set those to values :(

**Comment by jasonbahl** [comment]
@strootje so the way this is _supposed_ to be working is that when you make a query using `first` or `last` arguments, GraphQL uses WP_Query to over-query by 1 item. 

So: `{ posts( first: 10, after: "yourcursor"  ) { ...postsFields } }` will translate into a `WP_Query( 'posts_per_page' => 11 )` query

Then, if the number of items returned ( `count( $wp_query->posts )` ) is more than the number requested (`first: 10`) then `hasNextPage` is set to true. 

Same with last/before. 

The Relay spec for cursors states that `hasPreviousPage` should only return true when the query specifies a `before` cursor,  and `hasNextPage` should only be true if there's an `after` cursor.

So, if you make a general query like: 

```
posts {
  pageInfo {
    hasPreviousPage
    hasNextPage
  }
  edges {
    node {
        id
        title
     }
  }
}
```

Both `hasPreviousPage` and `hasNextPage` would be false, because there hasn't been `first` or `last` specified, so it's not clear if you're attempting to paginate forward or backward. 

If you specify `first: xx` or `last: xx` in your query, you _should_ get `hasNextPage` or `hasPreviousPage` set properly. 

I do think there are some bugs around pagination right now, but wanted to just share some info on how it _should_ work.

**Comment by strootje** [comment]
@jasonbahl Thanks for the explanation. I didn't know about the Relay spec.

**Comment by jasonbahl** [comment]
@izzygld wanted to follow up and see if this is still an issue or if we can close this?

**Comment by izzygld** [comment]
all is good.

**Comment by MoOx** [comment]
I am new to graphql and faced the "hasPreviousPage is always false omg" (I am using first/after).
I am generating a static website and would like to generate only /after/:cursor pages. Is there a way to know if we have a previous page when using first/after params only?
If my question is not asked at the right place, please point me the right direction as I don't really know how to ask

**Comment by jasonbahl** [comment]
Also, very flattered to have @MoOx using this plugin. . .Love what you do in the OSS community. 😄

**Comment by MoOx** [comment]
I am rewriting the front end of the website of my wife (https://github.com/MoOx/dame.bio) and would like to offer a "standard" pagination, with "older posts" ... and "newer posts" as soon as you are and an "older posts" page. I was thinking I could create links with only after/:cursor and still offer a "newer posts" button in order to paginate backwards. My brain is maybe not ready yet for graphql pagination spirit :D 
Is first/offset a combo planned for this wordpress (awesome!) plugin? I see the offset thing on 
https://graphql.org/learn/pagination/#pagination-and-edges

**Comment by MoOx** [comment]
Am I interpreting the spec correctly here by expecting hasPreviousPage to be set when using after?

https://facebook.github.io/relay/graphql/connections.htm#HasPreviousPage()

> HasPreviousPage(allEdges, before, after, first, last)
> ...
> If after is set:
>   If the server can efficiently determine that elements exist prior to after, return true.

**Comment by jasonbahl** [comment]
@MoOx you are indeed. I'll have to look into that. 🤔 

At the moment, I think we chose not to support that because it was easier not to, not necessarily because it's not efficient. I'll open a new issue to address that specifically and see what we can come up with. 

In the mean time, if you're using Apollo Client or another caching client, you _may_ be able to utilize the client side cache to calculate whether to show/hide the next/prev buttons. 🤔 

Thanks for pointing this out.

**Comment by jasonbahl** [comment]
@MoOx I created #594 to tackle this. Thanks!
